### ISSUE TITLE
structured logging: handle unusual keys?

### ISSUE THREAD

**Issue by pohly** [kubernetes/klog#275]
/kind feature

**Describe the solution you'd like**

The current design decision is to dump keys as strings without any checks. If those keys contain spaces, line breaks, or other weird characters, then the log output will get messed up.

Presumably this was done for the sake of performance. We should measure the impact of validating keys and substituting unsafe ones. If the overhead is acceptable, we might decide to make the code more robust.

TODO:
- [ ] add test cases with non-string keys (int, bool, struct) to `test/output.go` with the expected result for the current code, submit PR for discussion
- [ ] decide if and how to enhance the output
- [ ] if we agree on changes, implement those and compare performance of `go test -bench=.` before and after the changes

**Anything else you would like to add:**

Brought up in https://github.com/kubernetes/klog/pull/273#discussion_r758800965

/wg structured-logging

**Comment by harshanarayana** [comment]
@pohly  I was adding a few test cases to the `output.go` as suggested in the description of this issue and add some additional thing to benchmarking. But `zapr` has a different way of handling non string keys in the `KVList`. 

https://github.com/go-logr/zapr/blob/df10f4788c061a53564cebafc055fd9ad3f3255e/zapr.go#L156-L162

```golang
		// process a key-value pair,
		// ensuring that the key is a string
		key, val := args[i], args[i+1]
		keyStr, isString := key.(string)
		if !isString {
			// if the key isn't a string, DPanic and stop logging
			if zl.panicMessages {
				zl.l.WithOptions(zap.AddCallerSkip(1)).DPanic("non-string key argument passed to logging, ignoring all later arguments", zapIt("invalid key", key))
			}
			break
		}
```

I was thinking of adding a new attribute called `NonStringKeysSupported` in the `OutputConfig` and conditionally run the test cases based on it for the items in `output.go` Is there any other suggestion on handling this better ?

**Comment by pohly** [comment]
Not sure where this use comment is now, but it seems relevant, so let me quote and reply:
> I was adding a few test cases to the output.go as suggested in the description of this issue and add some additional thing to benchmarking. But zapr has a different way of handling non string keys in
 the KVList.: `DPanic("non-string key argument passed to logging, ignoring all later arguments", zapIt("invalid
 key", key)`
 
 So it panics instead of continuing with the tests?
 
 Skipping such tests is one way of avoiding this, but ultimately we want to test also such behavior. Perhaps the logger invocation can be wrapped with a `defer func() { if r := recover(); r != nil { ...` that then records the panic and continues? The panic then needs to be part of the expected log output.

**Comment by harshanarayana** [comment]
@pohly I was adding that comment from my phone. Seems like I deleted it by mistake. Thanks for getting back to me. 

> So it panics instead of continuing with the tests?

No it does not. It causes a log entry with `"level":"dpanic"` to show up. Part of how the `zap` handles things in development logger mode. 

I started adding a few tests and there are considerable difference in how the `zapr` handles the log in comparison with `klog`. Will open a PR with some example tests cases to start the discussion
